### PR TITLE
joystick: XInput. Use XInputGetCapabilitiesEx instead of fragile GuessXInputDevice

### PR DESCRIPTION
**Pull request by DJm00n**
XInputGetCapabilitiesEx (ordinal 108) is available in XInput 1.4 that is shipped with Windows 8 and newer.

I did some checks and it works reliable.

- Wired Xbox 360 Controller
image

- Wireless Xbox 360 Controller
image

- Xbox One Wired Connection
image

- Xbox One Wireless Connection
image

- Xbox Elite 2 Wired Connection
image

- Xbox Elite 2 BLE Connection
image

- Xbox Elite 2 Wireless Connection
image

### CODE REVIEW

File: src/joystick/windows/SDL_xinputjoystick.c [hunk 1]
```diff
@@ -287,8 +196,6 @@ static void AddXInputDevice(Uint8 userid, BYTE SubType, JoyStick_DeviceData **pC
     }
     (void)SDL_snprintf(pNewJoystick->path, sizeof(pNewJoystick->path), "XInput#%u", userid);
     if (!SDL_XInputUseOldJoystickMapping()) {
-        GuessXInputDevice(userid, &vendor, &product, &version);
-
         pNewJoystick->guid = SDL_CreateJoystickGUID(SDL_HARDWARE_BUS_USB, vendor, product, version, pNewJoystick->joystickname, 'x', SubType);
```
Comment (slouken) [inline src/joystick/windows/SDL_xinputjoystick.c:199]: This should probably use the Bluetooth bus for wireless gamepads.

File: src/joystick/windows/SDL_xinputjoystick.c [hunk 2]
```diff
@@ -114,116 +113,25 @@ static const char *GetXInputName(const Uint8 userid, BYTE SubType)
...
-                SDL_free(devices);
-                return;
-            }
-        }
-    }
-    SDL_free(devices);
-#endif /* !__WINRT__ */
+    *pVID = capabilities.VendorId;
+    *pPID = capabilities.ProductId;
+    *pVersion = capabilities.ProductVersion;
```
Comment (slouken) [inline src/joystick/windows/SDL_xinputjoystick.c:132]: Now that we can match on VID/PID, we should use raw input to look up the device name

File: src/joystick/windows/SDL_xinputjoystick.c [hunk 3]
```diff
@@ -114,116 +113,25 @@ static const char *GetXInputName(const Uint8 userid, BYTE SubType)
...
-                    /* We already have this device in our XInput device list */
-                    continue;
-                }
-
-                /* We don't actually know if this is the right device for this
-                 * userid, but we'll record it so we'll at least be consistent
-                 * when the raw device list changes.
-                 */
-                if (rdi.hid.dwVendorId == USB_VENDOR_VALVE &&
-                    rdi.hid.dwProductId == USB_PRODUCT_STEAM_VIRTUAL_GAMEPAD) {
```
Comment (slouken) [inline src/joystick/windows/SDL_xinputjoystick.c:198]: Note to self: Steam needs to be updated to hook XInputGetCapabilitiesEx() and return the correct VID/PID here.

**Comment by DJm00n** [comment]
Related link: https://github.com/Nemirtingas/OpenXinput/tree/OpenXinput1_4

File: src/joystick/windows/SDL_xinputjoystick.c (hunk 2 above)
Comment (slouken) [reply src/joystick/windows/SDL_xinputjoystick.c:132]: Can you make these changes so we can cherry pick it as a single commit to SDL2?

File: src/joystick/windows/SDL_xinputjoystick.c [hunk 4]
```diff
@@ -114,116 +113,25 @@ static const char *GetXInputName(const Uint8 userid, BYTE SubType)
...
-                    *pVID = (Uint16)rdi.hid.dwVendorId;
-                    *pPID = (Uint16)rdi.hid.dwProductId;
-                    *pVersion = (Uint16)rdi.hid.dwVersionNumber;
-                    SDL_free(devices);
-                    return;
-                }
-            }
-        }
+    /* Fixup for Wireless Xbox 360 Controller */
+    if (capabilities.ProductId == 0 && capabilities.Capabilities.Flags & XINPUT_CAPS_WIRELESS) {
```
Comment (DJm00n) [inline src/joystick/windows/SDL_xinputjoystick.c:125]: https://github.com/Nemirtingas/OpenXinput/blob/OpenXinput1_4/src/OpenXinput.cpp#L907-L916

**Comment by slouken** [comment]
Actually, I'll go ahead and merge this, and we can add name lookup afterwards.

Thanks!

File: src/joystick/windows/SDL_xinputjoystick.c (hunk 2 above)
Comment (slouken) [reply src/joystick/windows/SDL_xinputjoystick.c:132]: The VID/PID for all Xbox One controllers appears to be 0x045e/0x02ff, which has the generic controller name, so using raw input to look up the name isn't helpful. Never mind! :)

File: src/joystick/windows/SDL_xinputjoystick.c (hunk 1 above)
Comment (slouken) [reply src/joystick/windows/SDL_xinputjoystick.c:199]: Actually, this uses the USB bus for the wireless dongle, never mind.

**Pull request merged**